webdriver/protocol.py

```diff
@@ -0,0 +1,130 @@
+import contextlib
+import httplib
+import json
+import pytest
+import types
+
+import webdriver
+
+
+class HTTPRequest(object):
+    def __init__(self, host, port):
+        self.host = host
+        self.port = port
+
+    def head(self, path):
+        return self._request("HEAD", path)
+
+    def get(self, path):
+        return self._request("GET", path)
+
+    @contextlib.contextmanager
+    def _request(self, method, path):
+        conn = httplib.HTTPConnection(self.host, self.port)
+        try:
+            conn.request(method, path)
+            yield conn.getresponse()
+        finally:
+            conn.close()
+
+
+@pytest.fixture
+def http(request, session):
+    return HTTPRequest(session.transport.host, session.transport.port)
+
+
+@pytest.fixture
+def new_session(session):
+    session.end()
+    assert session.session_id is None
+    return session
+
+
+def test_provides_http(http):
+    # 4.0
+    with http.head("/") as resp:
+        assert resp.status == 404
+
+
+def test_send_an_error(http):
+    # 4.3 send an error, steps 1-5
+    with http.get("/cream") as resp:
+        assert resp.status == 404
+        body = json.load(resp)
+    assert "error" in body
+    assert "message" in body
+    assert "stacktrace" in body
+    assert isinstance(body["error"], types.StringTypes)
+    assert isinstance(body["message"], types.StringTypes)
+    assert isinstance(body["stacktrace"], types.StringTypes)
+    assert body["error"] == "unknown command"
+
+
+def test_unknown_command(http, new_session):
+    # 4.3 step 3
+    with http.head("/foo") as resp:
+        assert resp.status == 404
+    with http.get("/foo") as resp:
+        assert resp.status == 404
+
+    # jump to step 1
+    new_session.url
+
+
+def test_invalid_session_id(http, new_session):
```

Comment:
Guessing this is invalid because the session ID is not a UUID? https://w3c.github.io/webdriver/webdriver-spec.html#dfn-session-id Or is it because there's no session that has been created with that ID? The test name doesn't make this clear.